### ISSUE TITLE
refresh_tokens may be removed from storage without returning the new tokens (access, id, refresh)

### ISSUE THREAD

**Issue by adimkov** [DuendeSoftware/Support#667]
**Which version of Duende IdentityServer are you using?**
Duende.IdentityServer 6.2.3
**Which version of .NET are you using?**
NetCore 7

**Describe the bug**
In custom implementation of `ClaimsService` could happen exceptions (which is normal because of dependency). However, it leads to the refresh_token removal from storage without returning a new one. The root of the error `TokenResponseGenerator` is because the refersh_token [removed](https://github.com/DuendeSoftware/IdentityServer/blob/274b79fa3159dda3c530418956ddb9d2dc0c5dd3/src/IdentityServer/ResponseHandling/Default/TokenResponseGenerator.cs#LL220C56-L220C56) before generating the [idToken](https://github.com/DuendeSoftware/IdentityServer/blob/274b79fa3159dda3c530418956ddb9d2dc0c5dd3/src/IdentityServer/ResponseHandling/Default/TokenResponseGenerator.cs#L229)
![image](https://github.com/DuendeSoftware/Support/assets/1008064/a6686846-b49d-4fc3-bfc6-23c6307df88a)

This issue prevents integrators from refreshing access_token after an unsuccessful previous operation. 

**To Reproduce**
configure client to use "one time" refresh tokens

1) Implement custom `ClaimsService`, which will throw an error
2) try to refresh token

**Expected behavior**
1) Error should be returned
2) refresh_token should not be removed

In other words, this operation should be atomic.

**Log output/exception with stacktrace**

```
data
```

**Additional context**

**Comment by josephdecock** [comment]
Thanks for this feedback - making this operation atomic is an interesting idea that we might consider for a future release. We probably can't make that change outside a major release though, because we would most likely have to change the underlying store abstractions, which would be a breaking change.

Apart from server side errors, one time use refresh tokens are unreliable when there are network outages - if your identity server host receives the rotation request, but your client doesn't receive the response due to a bad network connection, the token becomes unusable in that case as well. The solution to this problem is to customize the RefreahTokenService - https://docs.duendesoftware.com/identityserver/v6/tokens/refresh/#accepting-consumed-tokens. Similar customizations could work around transient server side errors during the refresh process.

Finally, I would note that you can safely switch to reusable refresh tokens if your clients are confidential clients (aka, they have a client secret/are not a spa or native app).

**Comment by adimkov** [comment]
Hey [Joe](https://github.com/josephdecock),

Thank you for your reply and proposing solutions. We will mull over the option of a custom implementation of RefreshTokenService. However, it looks overcomplicated and has too many assumptions, like "time when the token is valid." The issue remains if an integrator will not refresh the token within a specified time. Also, the background job must go through all consumed tokens to remove obsolete ones. This reduces storage options for tokens, and in our case, it's critical.

 We have now fixed it by overriding `ProcessRefreshTokenRequestAsync` of `TokenResponseGenerator` and moved `CreateIdTokenFromRefreshTokenRequestAsync` before `RefreshTokenService.UpdateRefreshTokenAsync`. The fix is simple and works for us.


Thank you,
Anton
